[sosreport] Allow policies to change report name generation

TurboTurtle opened the pull request:
Previously, sosreport tarball names consisted of a very long date
string that was not pleasant to read, especially when dealing with
multiple sosreports for an ongoing issue.

This allows policies to determine how the report name should be
formatted by adding a 'name_scheme' attribute to the policies. By
default, this is set to 'legacy' and will result in names that are
unchanged - E.G this will give tarball names like
'sosreport-tux.123456-20170120185433'.

Additionally, there is a new 'friendly' name scheme that results in
tarball names like 'sosreport-tux-123456-2017-12-24-ezcfcop.tar.xz'.

A policy may also set name_scheme to 'custom' and provide it's own
pattern by setting the 'name_scheme_string' to a format() parsable
string.

This also sets the Red Hat policy to use the new 'friendly' pattern.

Resolves: #469

Signed-off-by: Jake Hunsaker <jhunsake@redhat.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?

pmoravec commented on sos/policies/redhat.py:
```diff
@@ -45,6 +45,7 @@ class RedHatPolicy(LinuxPolicy):
     _in_container = False
     _host_sysroot = '/'
     default_scl_prefix = '/opt/rh'
+    name_scheme = 'friendly'
```
Is there a reason to have non-default `name_scheme` for one specific policy? I understand your intention, but I dont understand what is specific on the scheme for different policies.

(maybe have this as a user-configured option? or rather directly have sosreport option --report-namescheme="-{name}-{case}-%Y-%m-%d-{rand}" that will be appended after "`sosreport`" string?)

TurboTurtle replied:
I am wanting to change the RH naming scheme because that's where I have a pulse with people using sosreports on a daily basis. I can't say the same for any other OSes, so I left those alone.

I think this should be policy driven because it's unlikely (in my opinion) that a user would actively change the pattern via individual runs of sos because that would be a bit clunky.

pmoravec commented on sos/policies/__init__.py:
```diff
@@ -261,11 +265,40 @@ def get_archive_name(self):
+            nstr += '-%Y%m%d%H%M%S'
+        if self.name_scheme == 'friendly':
+            case = ('-' + case if case else '')
+            nstr += '-%Y-%m-%d-{rand}'
```
From user experience, I prefer having also time (not only date) in the final report name. It is better for identifying and sorting reports from the same machine generated several times during the same day (well this rather happens during my testing than during a support experience).

But that is my personal point of view only.

TurboTurtle replied:
From #469 I dropped the time from the sosreport in favor of the random 7-character string as I believe it is easier to read at a glance than a timestamp. Even a further formatted date string is a bit harsh on the eyes in my opinion. The 7-character string can be used as both a global identifier when dealing with a mass of sosreports, and an identifier for multiple sosreports from the same day.

pmoravec replied:
Both makes sense; "harsh on the eyes" can be minimized using `-%Y-%m-%d_%H-%M-%S` format, but I understand it is subjective.

iranzo replied:
I prefer as pmoravec++ suggested, but I'll also suggest to add hostname in the file, for cases reported against OSP, where you can get 3 controllers + several computes on the same case it will make a lot easier to differentiate them

pmoravec replied:
hostname definitely should be part of the name (how I could forget it?). There are numerous situations where one needs to work with sosreports from several systems inter-communicating and having a problem, and we should be able to distinguish among the reports easily.

pmoravec commented:
In general, I like it. Some comments in the commit, IMHO the sosreport (cmdline) option would be nice here.

TurboTurtle commented:
Made some changes:

- Simplified the toggle between 'legacy', 'friendly' and 'custom' name patterns
- Removed the user prompt for a name and always set the short hostname now
- Added a `--label` option to allow a user-provided string to be appended after the hostname (and `--name` maps over to this now)
